packages/@lwc/template-compiler/src/codegen/helpers.ts

```diff
@@ -6,19 +6,113 @@
  */
 import * as t from '../shared/estree';
 import { toPropertyName } from '../shared/utils';
-import { BaseElement, ChildNode, LWCDirectiveRenderMode, Node } from '../shared/types';
+import {
+    Attribute,
+    BaseElement,
+    ChildNode,
+    Element,
+    LWCDirectiveRenderMode,
+    Node,
+    Property,
+} from '../shared/types';
 import {
     isParentNode,
     isSlot,
     isForBlock,
     isBaseElement,
     isIf,
     isDynamicDirective,
+    isElement,
+    isText,
+    isComment,
 } from '../shared/ast';
 import { TEMPLATE_FUNCTION_NAME, TEMPLATE_PARAMS } from '../shared/constants';
 
 import CodeGen from './codegen';
 
+function serializeAttrs(element: Element): string {
+    /**
+     * 0: styleToken in existing class attr
+     * 1: styleToken for added class attr
+     * 2: styleToken as attr
+     */
+    const attrs: string[] = [];
+    let hasClassAttr = false;
+
+    const collector = ({ name, value }: Attribute | Property) => {
+        let v = (value as any).value;
+
+        if (name === 'class') {
+            hasClassAttr = true;
+            v += '${0}';
+        }
+        if (typeof v === 'string') {
+            attrs.push(`${name}="${v}"`);
+        } else {
+            attrs.push(name);
+        }
+    };
+
+    element.attributes.forEach(collector);
+    element.properties.forEach(collector);
+
+    return attrs.join(' ') + (hasClassAttr ? '${2}' : '${1}${2}');
+}
+
+function serializeChildren(children: ChildNode[]): string {
+    let html = '';
+
+    children.forEach((child) => {
+        if (isElement(child)) {
+            html += serializeStaticElement(child);
+        } else if (isText(child)) {
+            html += (child.value as any).value;
```

Comment:
This needs to be escaped; otherwise it's vulnerable to an XSS injection:

```html
<template>
    <h1>Hello &lt;div&gt;world&lt;/div&gt;</h1>
</template>
```

![Screen Shot 2022-04-12 at 1 38 45 PM](https://user-images.githubusercontent.com/283842/163049838-e7b40af9-da7d-4a76-814a-c8fc18a9f912.png)

`master` does not have the same vulnerability; here is a screenshot of `master` behaving correctly:

![Screen Shot 2022-04-12 at 1 39 13 PM](https://user-images.githubusercontent.com/283842/163049918-e5060d17-1b88-4a29-81b1-0255a31dceb0.png)